docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,568 @@
+    - The extra information (i.e., the vote extensions) that is now kept in the blockchain is not really
+        needed *at every height* for a late node to catch up.
+        - This information is only needed to be able to *propose* at the height the validator considers
+          itself as caught-up. If a validator is indeed late for height *h*, it is useless (although
+          correct) for it to call `PrepareProposal`, or `ExtendVote`, since the block is already decided.
+        - Moreover, some use cases require pretty sizeable vote extensions, which would result in an
+          important waste of space in the blockchain.
+
+- **Solution 2.** *Skip* propose *step in Tendermint algorithm*.
+
+    This solution consists in modifying the Tendermint algorithm to skip the *send proposal* step in
+    heights where the node does not have the required vote extensions to populate the call to
+    `PrepareProposal`. The main idea behind this is that it should only happen when the validator is late
+    and, therefore, up-to-date validators have already proposed (and decided) for that height.
+    A small variation of this solution is, rather than skipping the *send proposal* step, the validator
+    sends a special *empty* or *bottom* (⊥) proposal to signal other nodes that it is not ready to propose
+    at (any round of) the current height.
+
+    The appeal of this solution is its simplicity. A possible implementation does not need to extend
+    the data structures, or change the current catch-up mechanisms implemented in the blocksync or
+    in the consensus reactor. When we lack the needed information (vote extensions), we simply rely
+    on another correct validator to propose a valid block in other rounds of the current height.
+
+    However, this solution can be attacked by a byzantine node in the network in the following way.
+    Let us consider the following scenario:
+
+    - all validators in *valset<sub>h</sub>* send out precommit messages, with vote extensions,
+      for height *h*, round 1, roughly at the same time,
+    - all those precommit messages contain non-`nil` precommit votes, which vote for block *b*
+    - all validators in *valset<sub>h</sub>* then wait until they gather enough precommit
+      messages for height *h*, round 1, in order to decide in height *h*,
+    - all precommit messages sent in height *h*, round 1, and all messages sent in
+      height *h*, round *r > 1* get delayed indefinitely,
+    - an intermediate (malicious) full node *m* manages to receive block *b*, and gather more than
+      *2n<sub>h</sub>/3* precommit messages for height *h*, round 1,
+    - one way or another, the solution should have a mechanism for a full node to tell another full
+      node it is late; this mechanism should somehow include a message type containing － at the very
+      least － the decided block and a commit (not necessarily an extended commit) for *h*;
+      let us call this message type *catch-up message*,
```

Comment:
Yes. I know the example is very general, and so is the definition of catch-up message. The reason is we're just showing the weakness of this solution in abstract terms. Probably a more formal definition would be needed to show the problem will be there regardless of the particular implementation (but this can be done later).

Reply:
I have re-worked this part of the text. You might check out the new version